File: sassy-talk-clean/android-app/app/src/main/java/com/sassyconsulting/sassytalkie/license/LicensePromo.kt
// Copyright (c) 2026 Shane Smith / Sassy Consulting LLC. All rights reserved.
// Proprietary source. This notice is Copyright Management Information (17 U.S.C. 1202); removal or alteration prohibited.
// CodeMark: SCLLC1-sassytalkie-Q7BFTEJJ46J2
package com.sassyconsulting.sassytalkie.license

import android.content.Context
import android.util.Log
import com.sassyconsulting.sassytalkie.BuildConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Promo-code redemption against the relay worker's `/license/promo` endpoint.
 * Used by the Play paywall (friends & family bypass) and refresh logic.
 */
object LicensePromo {
    private const val TAG = "LicensePromo"
    private const val REFRESH_WHEN_REMAINING_SEC = 15L * 24 * 3600
    private val LICENSE_RE = Regex("^SASSY(?:-[A-HJ-NP-Z2-9]{5}){4}$")
    private val PROMO_RE = Regex("^[A-Z0-9-]{6,40}$")

    sealed class RedeemResult {
        data object Ok : RedeemResult()
        data object InvalidFormat : RedeemResult()
        data object NetworkError : RedeemResult()
        data class Rejected(val message: String) : RedeemResult()
    }

    fun hasValidReceipt(context: Context): Boolean {
        val p = LicenseStore.prefs(context) ?: return false
        if (p.getString(LicenseStore.KEY_KIND, null) != "promo") return false
        val exp = p.getLong(LicenseStore.KEY_RECEIPT_EXP, 0L)
        return exp > System.currentTimeMillis() / 1000
    }

    /** Mirrors worker [normalizePromo]: 6–40 chars A-Z 0-9 -, not a license key. */
    fun normalizeCode(raw: String): String? {
        val code = raw.trim().uppercase().replace(Regex("\\s+"), "")
        if (LICENSE_RE.matches(code)) return null
        return if (PROMO_RE.matches(code)) code else null
    }

    fun redeemBlocking(context: Context, rawCode: String): RedeemResult {
        val code = normalizeCode(rawCode) ?: return RedeemResult.InvalidFormat
        return when (val api = callPromo(context, code)) {
            is ApiResult.Ok -> {
                persist(context, code, api.expiresAt)
                RedeemResult.Ok
            }
            is ApiResult.Rejected -> RedeemResult.Rejected(api.message)
            ApiResult.NetworkError -> RedeemResult.NetworkError
        }
    }

    fun refreshIfNeeded(context: Context, onResult: (Boolean) -> Unit) {
        val p = LicenseStore.prefs(context) ?: return onResult(false)
        val code = p.getString(LicenseStore.KEY_LICENSE, null) ?: return onResult(false)
        if (p.getString(LicenseStore.KEY_KIND, null) != "promo") return onResult(false)
        val exp = p.getLong(LicenseStore.KEY_RECEIPT_EXP, 0L)
        val now = System.currentTimeMillis() / 1000
        if (exp <= now) return onResult(false)
        if (exp - now > REFRESH_WHEN_REMAINING_SEC) return onResult(true)

        CoroutineScope(Dispatchers.IO).launch {
            when (val res = callPromo(context, code)) {
                is ApiResult.Ok -> {
                    p.edit().putLong(LicenseStore.KEY_RECEIPT_EXP, res.expiresAt).apply()
                    onResult(true)
                }
                is ApiResult.Rejected -> {
                    Log.w(TAG, "Promo rejected on refresh: ${res.message}")
                    p.edit()
                        .remove(LicenseStore.KEY_RECEIPT_EXP)
                        .remove(LicenseStore.KEY_KIND)
                        .remove(LicenseStore.KEY_LICENSE)
                        .putBoolean(LicenseStore.KEY_UNLOCKED, false)
                        .apply()
                    onResult(false)
                }
                ApiResult.NetworkError -> onResult(exp > now)
            }
        }
    }

    private fun persist(context: Context, code: String, expiresAt: Long) {
        // Promo entitlement is RECEIPT-DRIVEN (time-limited): store only the
        // expiry + kind, and let isUnlockedCached honor it via hasValidReceipt.
        // Never set the permanent KEY_UNLOCKED flag here — that flag is checked
        // first and would make a 30-day promo last forever.
        LicenseStore.prefs(context)?.edit()
            ?.putString(LicenseStore.KEY_LICENSE, code)
            ?.putString(LicenseStore.KEY_KIND, "promo")
            ?.putLong(LicenseStore.KEY_RECEIPT_EXP, expiresAt)
            ?.apply()
    }

    private sealed class ApiResult {
        class Ok(val expiresAt: Long) : ApiResult()
        class Rejected(val message: String) : ApiResult()
        data object NetworkError : ApiResult()
    }

    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private fun callPromo(context: Context, code: String): ApiResult {
        return try {
            val body = JSONObject()
                .put("code", code)
                .put("device_id", deviceId(context))
                .put("device_name", android.os.Build.MODEL)
                .put("app_version", BuildConfig.VERSION_NAME)
                .toString()
                .toRequestBody("application/json".toMediaType())
            val req = Request.Builder()
                .url(BuildConfig.LICENSE_API_BASE + "/license/promo")
                .post(body)
                .build()
            http.newCall(req).execute().use { resp ->
                val json = JSONObject(resp.body?.string() ?: "{}")
                when {
                    resp.isSuccessful && json.optBoolean("ok") ->
                        ApiResult.Ok(json.optLong("expires_at"))
                    resp.code >= 500 -> ApiResult.NetworkError
                    else -> ApiResult.Rejected(json.optString("error", "Promo rejected"))
                }
            }
        } catch (e: Exception) {
            Log.w(TAG, "promo call failed: ${e.message}")
            ApiResult.NetworkError
        }
    }

    private fun deviceId(context: Context): String = LicenseStore.deviceId(context)
}

File: ios-native/LicensePromo.swift
// Copyright (c) 2026 Shane Smith / Sassy Consulting LLC. All rights reserved.
// Proprietary source. This notice is Copyright Management Information (17 U.S.C. 1202); removal or alteration prohibited.
// CodeMark: SCLLC1-sassytalkie-IOSPROMORED9M
//
//  LicensePromo.swift
//  Promo-code redemption against the relay `/license/promo` endpoint
//  (parity with Android LicensePromo.kt). Receipt is time-limited; do not
//  set the permanent StoreKit unlocked flag.

import Foundation
import UIKit

enum LicensePromo {
    static let licenseApiBase = "https://relay.sassyconsultingllc.com"
    private static let refreshWhenRemainingSec: TimeInterval = 15 * 24 * 3600
    private static let licenseRe = try! NSRegularExpression(
        pattern: "^SASSY(?:-[A-HJ-NP-Z2-9]{5}){4}$")
    private static let promoRe = try! NSRegularExpression(
        pattern: "^[A-Z0-9-]{6,40}$")

    private static let keyLicense = "promo_license_code"
    private static let keyKind = "promo_credential_kind"
    private static let keyReceiptExp = "promo_receipt_exp"
    private static let keyInstallId = "promo_install_id"

    enum RedeemResult {
        case ok
        case invalidFormat
        case networkError
        case rejected(String)
    }

    /// True when the cached credential kind is promo (may be expired).
    static var isPromoCredential: Bool {
        UserDefaults.standard.string(forKey: keyKind) == "promo"
    }

    static func hasValidReceipt() -> Bool {
        guard isPromoCredential else { return false }
        let exp = UserDefaults.standard.double(forKey: keyReceiptExp)
        return exp > Date().timeIntervalSince1970
    }

    /// Mirrors worker normalizePromo: 6–40 chars A-Z 0-9 -, not a license key.
    static func normalizeCode(_ raw: String) -> String? {
        let code = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            .uppercased()
            .replacingOccurrences(of: "\\s+", with: "", options: .regularExpression)
        let range = NSRange(code.startIndex..., in: code)
        if licenseRe.firstMatch(in: code, range: range) != nil { return nil }
        guard promoRe.firstMatch(in: code, range: range) != nil else { return nil }
        return code
    }

    static func redeemBlocking(_ rawCode: String) -> RedeemResult {
        guard let code = normalizeCode(rawCode) else { return .invalidFormat }
        switch callPromo(code) {
        case .ok(let expiresAt):
            persist(code: code, expiresAt: expiresAt)
            return .ok
        case .rejected(let message):
            return .rejected(message)
        case .networkError:
            return .networkError
        }
    }

    static func refreshIfNeeded(completion: @escaping (Bool) -> Void) {
        let defaults = UserDefaults.standard
        guard let code = defaults.string(forKey: keyLicense),
              defaults.string(forKey: keyKind) == "promo" else {
            completion(false)
            return
        }
        let exp = defaults.double(forKey: keyReceiptExp)
        let now = Date().timeIntervalSince1970
        if exp <= now {
            completion(false)
            return
        }
        if exp - now > refreshWhenRemainingSec {
            completion(true)
            return
        }
        DispatchQueue.global(qos: .utility).async {
            switch callPromo(code) {
            case .ok(let expiresAt):
                defaults.set(expiresAt, forKey: keyReceiptExp)
                completion(true)
            case .rejected:
                defaults.removeObject(forKey: keyReceiptExp)
                defaults.removeObject(forKey: keyKind)
                defaults.removeObject(forKey: keyLicense)
                completion(false)
            case .networkError:
                completion(exp > now)
            }
        }
    }

    private static func persist(code: String, expiresAt: Double) {
        // Receipt-driven only — never set StoreKitEntitlements.persistUnlocked.
        let defaults = UserDefaults.standard
        defaults.set(code, forKey: keyLicense)
        defaults.set("promo", forKey: keyKind)
        defaults.set(expiresAt, forKey: keyReceiptExp)
    }

    private enum ApiResult {
        case ok(Double)
        case rejected(String)
        case networkError
    }

    private static func callPromo(_ code: String) -> ApiResult {
        guard let url = URL(string: "\(licenseApiBase)/license/promo") else {
            return .networkError
        }
        let body: [String: String] = [
            "code": code,
            "device_id": deviceId(),
            "device_name": UIDevice.current.model,
            "app_version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: body) else {
            return .networkError
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        req.timeoutInterval = 10

        let sem = DispatchSemaphore(value: 0)
        var result: ApiResult = .networkError
        PinnedURLSession.shared.dataTask(with: req) { data, response, error in
            defer { sem.signal() }
            if error != nil {
                result = .networkError
                return
            }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            let json = (data.flatMap {
                try? JSONSerialization.jsonObject(with: $0) as? [String: Any]
            }) ?? [:]
            if (200..<300).contains(code), json["ok"] as? Bool == true {
                let exp = (json["expires_at"] as? Double)
                    ?? Double(json["expires_at"] as? Int ?? 0)
                result = .ok(exp)
            } else if code >= 500 {
                result = .networkError
            } else {
                let msg = (json["error"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                    ?? "Promo rejected"
                result = .rejected(msg)
            }
        }.resume()
        _ = sem.wait(timeout: .now() + 12)
        return result
    }

    private static func deviceId() -> String {
        if let id = UIDevice.current.identifierForVendor?.uuidString, !id.isEmpty {
            return id
        }
        if let existing = UserDefaults.standard.string(forKey: keyInstallId), !existing.isEmpty {
            return existing
        }
        let fresh = UUID().uuidString
        UserDefaults.standard.set(fresh, forKey: keyInstallId)
        return fresh
    }
}
